File: 0724-find-pivot-index/0724-find-pivot-index.cpp
class Solution {
public:
    int pivotIndex(vector<int>& nums) {

        int n = nums.size();

        if (n == 0) return -1;
        if (n == 1) return 0;

        // Step 1: Build prefix sum in-place
        for (int i = 1; i < n; i++) {
            nums[i] += nums[i - 1];
        }

        int total = nums[n - 1];

        // Step 2: Check pivot index
        for (int i = 0; i < n; i++) {

            int leftSum;
            if (i == 0)
                leftSum = 0;
            else
                leftSum = nums[i - 1];

            int rightSum = total - nums[i];

            if (leftSum == rightSum)
                return i;
        }

        return -1;
    }
};

File: 0724-find-pivot-index/0724-find-pivot-index.py
class Solution:
    def pivotIndex(self, nums: list[int]) -> int:
        # Calculate the total sum of all the numbers in the array
        total_sum = sum(nums)
        
        # Initialize left sum to 0 as there are no elements to the left of the first index initially
        left_sum = 0
        
        # Iterate through each index and its corresponding number in the array
        for i, num in enumerate(nums):
            # The right sum is the total sum minus the left sum and the current number
            if left_sum == total_sum - left_sum - num:
                return i
            
            # Add the current number to the left sum for the next iteration
            left_sum += num
            
        # Return -1 if no such pivot index exists
        return -1